src/oauthhandler.ts

```diff
@@ -0,0 +1,126 @@
+const URL_TOKEN = "https://discordapp.com/api/oauth2/token";
+const URL_REVOKE = "https://discordapp.com/api/oauth2/token/revoke";
+const URL_IDENTITY = "https://discordapp.com/api/users/%40me";
+
+const HTTP_UNAUTHORISED = 401;
+const HTTP_INTERNAL_ERROR = 500;
+
+interface IDiscordTokenResponse {
+    access_token: string;
+    expires_in: number;
+    refresh_token: string;
+    scope: string;
+    token_type: string;
+}
+
+/**
+ * This class handles OAuth identification for matching
+ * Discord users to Matrix accounts.
+ */
+export class OAuthHandler {
+    private stateUserMap: Map<string, string>;
+    constructor(private config: DiscordBridgeConfigAuth, private bridge: Bridge, private store: DiscordStore) {
+        this.stateUserMap = new Map();
+    }
+
+    public bindEndpoint() {
+        this.bridge.addAppServicePath({
+            handler: (req: Request, res: Response) => {
+                this.onGetOAuth(req, res).catch((ex) => {
+                    log.error("Failure during oauth:", ex);
+                    if (res.finished) {
+                        return;
+                    }
+                    res.sendStatus(HTTP_INTERNAL_ERROR);
+                });
+            },
+            method: "GET",
+            path: "/_bridge/oauth",
+        });
+        log.info("Bound _bridge/oauth for oAuth");
+    }
+
+    public handleOAuthRequest(userid: string) {
+        // 1. Get the userId
+        const stateKey = uuid();
+        this.stateUserMap.set(stateKey, userid);
+        const url = `${this.authUrl}&state=${stateKey}`;
+        return url;
+    }
+
+    public async onGetOAuth(req: Request, res: Response) {
+        const userId = this.stateUserMap.get(req.query.state);
+        if (!userId) {
+            log.warn("Ignorning oauth request for unknown user");
+            res.status(HTTP_UNAUTHORISED).send("Error: User not found");
+            return;
+        }
+        this.stateUserMap.delete(req.query.state);
+        log.info("Got OAuth Request for", userId);
+        const formData = {
+            client_id: this.config.clientID,
+            client_secret: this.config.clientSecret,
+            code: req.query.code,
+            grant_type: "authorization_code",
+            redirect_uri: this.config.oAuthUrl,
+            scope: this.scopes(),
+        };
+        try {
+            // Tslint fails to recognise these as promises.
+            /* tslint:disable-next-line await-promise */
+            const atRes = JSON.parse(await (request.post({
+                formData,
+                simple: true,
+                url: URL_TOKEN,
+            }))) as IDiscordTokenResponse;
+            const dataObj = new DbAccessToken();
+            dataObj.DiscordId = await this.getIdentity(atRes.access_token);
+            dataObj.AccessToken = atRes.access_token;
+            dataObj.RefreshToken = atRes.refresh_token;
+            dataObj.ExpiresIn = atRes.expires_in || 1;
```

Comment:
what do we do when the token expires?

Reply:
Nothing. I guess we could invalidate the mapping but that seems kinda rude and I don't think it's a good idea to ask users to reauth every few weeks to keep their mappings. Perhaps we could just require reauth for any *new* explicit actions (if I can think of any) and anything automatic like power levels will continue to happen.